clash-prelude/src/Clash/Class/Finite/Internal.hs

```diff
@@ -0,0 +1,967 @@
+  gLowestMaybe  = M1 <$> gLowestMaybe
+  gHighest      = M1 gHighest
+  gHighestMaybe = M1 <$> gHighestMaybe
+  gPredMaybe    = fmap M1 . gPredMaybe . unM1
+  gSuccMaybe    = fmap M1 . gSuccMaybe . unM1
+  gIth          = M1 . gIth
+  gIndex        = gIndex . unM1
+
+instance (GFinite a, GFinite b) => GFinite (a :*: b) where
+  type GElementCount (a :*: b) = GElementCount a * GElementCount b
+  gElements = concatMap (\a -> (a :*:) <$> gElements @b) (gElements @a)
+
+  gLowest
+    | Dict <- mulPositiveImpliesPositiveOperands @(GElementCount a) @(GElementCount b)
+    = gLowest :*: gLowest
+  gLowestMaybe = (:*:) <$> gLowestMaybe  <*> gLowestMaybe
+
+  gHighest
+    | Dict <- mulPositiveImpliesPositiveOperands @(GElementCount a) @(GElementCount b)
+    = gHighest :*: gHighest
+  gHighestMaybe = (:*:) <$> gHighestMaybe <*> gHighestMaybe
+
+  gPredMaybe (a :*: b) =
+        (:*:)             a <$> gPredMaybe b
+    <|> (:*:) <$> gPredMaybe a <*> gHighestMaybe
+  gSuccMaybe (a :*: b) =
+        (:*:)            a <$> gSuccMaybe b
+    <|> (:*:) <$> gSuccMaybe a <*> gLowestMaybe
+
+  gIth x = gIth (resize $ x `div` m) :*: gIth (resize $ x `mod` m)
+   where
+    m = natToNum @(GElementCount b)
+
+  gIndex (a :*: b) =
+      resize (gIndex a) * natToNum @(GElementCount b)
+    + resize (gIndex b)
+
+instance (GFinite a, GFinite b) => GFinite (a :+: b) where
+  type GElementCount (a :+: b) = GElementCount a + GElementCount b
+  gElements = (L1 <$> gElements @a) ++ (R1 <$> gElements @b)
+
+  gLowest = case compareSNat (SNat @1) (SNat @(GElementCount a)) of
+    SNatLE -> L1 gLowest
+    SNatGT -> case compareSNat (SNat @1) (SNat @(GElementCount b)) of
+      SNatLE -> R1 gLowest
+      SNatGT -> case zeroLeAdd @(GElementCount a) @1 of
+        Dict -> case zeroLeAdd @(GElementCount b) @1 of {}
+  gLowestMaybe = L1 <$> gLowestMaybe  @a <|> R1 <$> gLowestMaybe  @b
+
+  gHighest = case compareSNat (SNat @1) (SNat @(GElementCount b)) of
+    SNatLE -> R1 gHighest
+    SNatGT -> case compareSNat (SNat @1) (SNat @(GElementCount a)) of
+      SNatLE -> L1 gHighest
+      SNatGT -> case zeroLeAdd @(GElementCount a) @1 of
+        Dict -> case zeroLeAdd @(GElementCount b) @1 of {}
+  gHighestMaybe = R1 <$> gHighestMaybe @b <|> L1 <$> gHighestMaybe @a
+
+  gPredMaybe = \case
+    L1 x -> L1 <$> gPredMaybe x
+    R1 x -> R1 <$> gPredMaybe x <|> L1 <$> gHighestMaybe
+
+  gSuccMaybe = \case
+    R1 x -> R1 <$> gSuccMaybe x
+    L1 x -> L1 <$> gSuccMaybe x <|> R1 <$> gLowestMaybe
+
+  gIth x
+    | e x < n   = L1 $ gIth $ truncateB x
+    | otherwise = R1 $ gIth $ truncateB $ e x - n
+   where
+    n = natToNum @(GElementCount a)
+    e = extend @Index @(GElementCount (a :+: b))
+      @(If (GElementCount (a :+: b) <=? GElementCount a) 1 0)
+
+  gIndex = \case
+    L1 x -> extend (gIndex x)
+    R1 x -> extend (gIndex x) + natToNum @(GElementCount a)
+
+instance Finite Void
+instance Finite ()
+instance Finite Bool
+instance Finite Ordering
+instance Finite (Proxy a)
```

Comment:
`SNat` is really just a `Proxy` with benefits, so just like we have the `Proxy` instance I feel we should have an `SNat` instance.

Same for `SSymbol`.